### PR TITLE
feat(Ts):  plugin templates

### PR DESCRIPTION
**Pull request by jaxzhou**
### ⚠️ &nbsp;&nbsp;Pre Checklist

> Please complete _ALL_ items in this checklist, and remove before submitting

- [ ] I have read through the [Contributing](../CONTRIBUTING.md) Documentation
- [x] This PR is using a `label` (bug, feature etc.)
- [x] My code is has necessary documentation (if appropriate)
- [x] I have added any relevant tests
- [ ] This section (**⚠️ &nbsp;&nbsp;Pre Checklist**) will be removed when submitting PR

# Summary

### Key Points

- [x] This is a breaking change
- [x] New or existing documentation is updated

### Description
we use plugin to manage package of tasks. and how task organized and how the datas/entities declares and migrates is the key points for next developments
This PR defined the core interfaces and decorators with jira plugin as a sample. that would help to understood how to write a plugin code.
We still need do some contribute on the framework, to make all things running. here is the framework and we would fill the missing parts(unimplement parts) lately 

### Does this close any open issues?
#184 

#### How plugin work?

1. Plugin Worker is triggered by Apps. Use queue producer to push a plugin task into queue with plugin name and data
```
producer.add('Jira', {host:'', username: '', password: ''});
```

2. Plugin is Loaded in Queue Consumer. the nestjs IOC framework would help to get a instance of Plugin
```
moduleRef.resolve('Jira', contextId, {strict: false});
```

3. Plugin resolve self with a Task DAG. 
 ```diff
- resolver not implement in this PR
- Plugin resolve the DAG by themself?  right now yes. we could change to another method in the future
```

4. A Task Service will read the Task DAG and start a Session to manage tasks execution
 ```diff
- Task service not in this PR
```

5. Task Service push task into queue service with task name and task data
```
producer.add('JiraIssueCollector', {host:'', username: '', password: ''});
```

6. Task Service recived the task completed events and check if has next task. the do step 5. if no task anymore finished the pipline
```
on('job:finished', (jobId, datas) => {  producer.add('JiraLeadtimeEnricher', datas); });
```

#### How plugin dependency described

- Plugin is the organizer to manage task/entity dependency.
```
@Injectable({ scope: Scope.TRANSIENT })
class Jira implements Plugin {
  constructor( private resolver: DependencyResolver) {}

  async execute(options: JiraOptions): Promise<DAG> {
    //TODO: Add jira collector and enrichment
    const dag = await this.resolver.resolve(IssueLeadTimeEntity);
    return dag;
  }
}
```

- Entities may have one Task to exports. a task is one of collector or enricher.

- The task would required some Entities, that should Import in Task

```
@Imports([RequiredEntity, AnotherReuqired.Entity])
@Exports(SampleEntity)
class SampleEnricher implements Task {
    async execute(): Promise<any> {
        //do enrich here
        return {};
    }
}
```

- Only one Entity should be Exports in Task

- Dependency Resolver would get the Task DAG for task exexution managment.

#### How to manage the Entities in Plugin

- One Task With The Exports Entities
   - task.ts
   - task.entity.ts
   - task.spec.ts

- All migrates should put at /plugins/YourPlugin/src/migrates, generator your migrate file with typeorm 
```
typeorm migrate:create -d /plugins/YourPlugin/src/migrates -n mymigrate
````

### CODE REVIEW

**Review by likyh — COMMENTED** [review]
LGTM

File: plugins/core/src/base.entity.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  PrimaryGeneratedColumn,
+  UpdateDateColumn,
+} from 'typeorm';
+
+class BaseEntity {
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
+
+  @CreateDateColumn()
+  createAt: Date;
```
Comment (narrowizard) [inline plugins/core/src/base.entity.ts:12]: A name of `createdAt` or `createTime` is better than `createAt`, right?

**Review by narrowizard — APPROVED** [review]
LGTM

File: apps/queue/src/consumer/index.ts [hunk 2]
```diff
@@ -1,15 +1,16 @@
 import { DynamicModule } from '@nestjs/common';
 import { BullQueueModule } from '../bull/queue.module';
-import { PluginModule } from './plugins';
+import PluginModule from 'plugins/core/src/plugins.module';
 import { ConsumerService } from './service';
+import Jira from 'plugins/jira/src';
```
Comment (klesh) [inline apps/queue/src/consumer/index.ts:5]: This doesn't seem right, Framework should not depend on a specific plugin, or it's not optional...

Comment (likyh) [reply apps/queue/src/consumer/index.ts:5]: It fix in my MR https://github.com/merico-dev/lake/pull/243

Comment (klesh) [reply apps/queue/src/consumer/index.ts:5]: I see, so we should merge this one first?

File: plugins/core/src/base.entity.ts [hunk 3]
```diff
@@ -1,18 +1,18 @@
 import {
-  CreateDateColumn,
+  CreatedDateColumn,
```
Comment (joncodo) [inline plugins/core/src/base.entity.ts:2]: @jaxzhou please review this commit. I can also make this as a PR to your branch if you want

File: apps/queue/src/consumer/service.ts [hunk 4]
```diff
@@ -2,7 +2,8 @@ import { InjectQueue } from '@nestjs/bull';
 import { Injectable } from '@nestjs/common';
 import { ModuleRef } from '@nestjs/core';
 import Bull, { Queue } from 'bull';
-import Plugin from 'plugins/core/src/plugin';
+import IExecutable from 'plugins/core/src/executable.interface';
+import { DAG } from 'plugins/core/src/dependency.resolver';
```
Comment (joncodo) [inline apps/queue/src/consumer/service.ts:6]: What does DAG stand for?

Comment (joncodo) [reply apps/queue/src/consumer/service.ts:6]: Directed Acyclic Graph

File: plugins/README.md [hunk 5]
```diff
@@ -1,34 +1,58 @@
...
+- 3. Plugin resolve self with a Task DAG.
+
+- 4. A Task Service will read the Task DAG and start a Session to manage tasks execution
+
+- 5. Task Service push task into queue service with task name and task data
+```
+producer.add('JiraIssueCollector', {host:'', username: '', password: ''});
+```
+
+- 6. Task Service recived the task completed events and check if has next task. the do step 5. if no task anymore finished the pipline
```
Comment (joncodo) [inline plugins/README.md:22]: It seems that "has next task" is not implemented here. Can you clarify?

File: apps/queue/src/consumer/service.ts (hunk 4 above)
Comment (jaxzhou) [reply apps/queue/src/consumer/service.ts:6]: yes, the tasks dependency would be a DAG.

File: plugins/README.md (hunk 5 above)
Comment (jaxzhou) [reply plugins/README.md:22]: this relative to #246 ,another PR will do some simple implements

File: plugins/core/src/base.entity.ts (hunk 3 above)
Comment (jaxzhou) [reply plugins/core/src/base.entity.ts]: great job.

File: apps/queue/src/consumer/index.ts (hunk 2 above)
Comment (jaxzhou) [reply apps/queue/src/consumer/index.ts:5]: this PR should merge first. maybe~

**Comment by hezyin** [comment]
LGTM

**Pull request merged**